### DISCUSSION TITLE
Spritesheet / Generalized texture atlas

### DISCUSSION THREAD

**Discussion by ffreyer** [JuliaPlots/Makie.jl#1590] [Ideas]
Currently it's possible to scatter an image, at least in GLMakie:
```julia
using GLMakie, FileIO
img = load(Makie.assetpath("cow.png"));
scatter(rand(10), marker = img, markersize = 40)
```
![Screenshot from 2022-01-14 15-46-31](https://user-images.githubusercontent.com/10947937/149534449-bbbb2c60-2992-4c92-a47d-067f279b91bf.png)

You can also use a vector of images to plot different images for each marker, which then ends up getting combined into one texture in GLMakie. (Either by using a GL_TEXTURE_2D_ARRAY or by packing images into a texture atlas.) This has a few problems:
1. Every image is assumed to be unique. If you want to scatter 10 different images to 1000 different positions you would either have to use 10 different scatter plots or create an array of 1000 images. The latter may then fail due to texture size constraints.
2. If you want to animate a sprite you currently need to either replace the image frequently or figure out some hacky way to generate a texture atlas and modify the uv's.
3. If you already have a spritesheet/texture atlas as one image there is no easy way for you to use it as one.

These problems could be solved by adding a user facing sprite sheet/texture atlas. Some things that would be required/useful:
- Creating a texture atlas from an image with
    - regular uv sizes spanning the whole image
    - regular uv sizes with line-lengths (i.e. [7, 3, 6] meaning 7 on the first line, 3 on the second, 6 on the third)
    - user given uv's
    - irregular uv's
- Creating a texture atlas from multiple images with automatic packing and uv's based on their size
- Creating an empty texture atlas of a given size
- pushing and appending images
- pushing and appending uv's
- integer indices to refer to images by insertion order

The plot signature could be `scatter(positions, marker = [1, 3, 7, 3, 4], atlas = my_texture_atlas)`.

It might also be useful to extend this to `meshscatter` for scattering meshes with varying textures more easily.

**Comment by SimonDanisch** [comment]
That would be nice. If we generalize the Texture atlas, we should think about putting it into its own package.
Also, I've never gotten around to use the "new" GuillotineBinPack I added to [Packing.jl](https://github.com/JuliaGeometry/Packing.jl) some time ago... Our current packing algorithm is super wasteful :D
